packages/components/src/confirm-dialog/component.tsx

```diff
@@ -0,0 +1,107 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import { forwardRef, useEffect, useState } from 'react';
+// eslint-disable-next-line no-restricted-imports
+import type { Ref, SyntheticEvent, KeyboardEvent } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Modal from '../modal';
+import type { OwnProps } from './types';
+import {
+	useContextSystem,
+	contextConnect,
+	WordPressComponentProps,
+} from '../ui/context';
+import { Flex } from '../flex';
+import Button from '../button';
+import * as styles from './styles';
+import { useCx } from '../utils/hooks/use-cx';
+
+function ConfirmDialog(
+	props: WordPressComponentProps< OwnProps, 'div', false >,
+	forwardedRef: Ref< any >
+) {
+	const {
+		isOpen,
+		message,
+		onConfirm,
+		onCancel = () => {},
+		...otherProps
+	} = useContextSystem( props, 'ConfirmDialog' );
+
+	const cx = useCx();
+	const wrapperClassName = cx( styles.wrapper );
+
+	const [ _isOpen, setIsOpen ] = useState< Boolean >();
+	const [ _selfClose, setSelfClose ] = useState< Boolean >();
+
+	useEffect( () => {
+		// We only allow the dialog to close itself if `isOpen` is *not* set.
+		// If `isOpen` is set, then it (probably) means it's controlled by a
+		// parent component. In that case, `selfClose` might do more harm than
+		// good, so we disable it.
+
+		const isIsOpenSet = typeof isOpen === 'boolean';
+		setIsOpen( isIsOpenSet ? isOpen : true );
+		setSelfClose( ! isIsOpenSet );
+	}, [ isOpen ] );
+
+	// @todo improve type, should handle keyboard and mousevent
+	const handleEvent = ( callback: ( event: SyntheticEvent ) => void ) => (
+		event: SyntheticEvent
+	) => {
+		callback( event );
+		if ( _selfClose ) {
+			setIsOpen( false );
+		}
+	};
+
+	const handleEnter = ( event: KeyboardEvent< HTMLDivElement > ) => {
+		if ( event.key === 'Enter' ) {
+			handleEvent( onConfirm )( event );
+		}
+	};
+
+	return (
+		<>
+			{ _isOpen && (
+				<Modal
+					title={ message }
+					overlayClassName={ wrapperClassName }
+					onRequestClose={ handleEvent( onCancel ) }
+					onKeyDown={ handleEnter }
+					closeButtonLabel={ __( 'Cancel' ) }
+					isDismissible={ true }
+					forwardedRef={ forwardedRef }
+					{ ...otherProps }
+				>
+					<Flex justify="flex-end">
+						<Button
+							variant="secondary"
+							onClick={ handleEvent( onCancel ) }
+						>
+							{ __( 'Cancel' ) }
```

Comment:
Do you think we should allow users of this component to change the `Cancel` and `OK` text? We could potentially expose them as `string` props, something like `confirmLabel` and `cancelLabel` ?

Reply:
The first version was meant to be a (mostly) 1:1 model of `confirm`, hence the hardcoded labels, but yep! @jasmussen also suggested it in his comment [above](https://github.com/WordPress/gutenberg/pull/34153#issuecomment-920661868).

Reply:
I'm considering working on that in a follow-up PR, and include a more detailed review / discussion of the visual design aspects. What do you think? cc @ciampo @jasmussen

Reply:
If the design still looks like https://github.com/WordPress/gutenberg/pull/34153#issuecomment-905112333, I think we need to at least replace the secondary style cancel button with a tertiary style.

The title wrapping two lines is a bit challenging. The default confirm dialog shows the "Would you like to privately publish this post now?" as just text, not a heading. Can we remove the heading unless explicitly set? I.e. something like:

<img width="503" alt="Screenshot 2021-09-25 at 11 18 48" src="https://user-images.githubusercontent.com/1204802/134766443-3f9c9e3c-3eba-4adb-a678-c2e425ce271b.png">

We do have a little wiggleroom to keep PRs small, but these would be nice changes to get along.

Reply:
@jasmussen That's great feedback, thanks! I'll include those changes as part of this changeset.